### DISCUSSION TITLE
Subclass UnicornView?

### DISCUSSION THREAD

**Discussion by nerdoc** [django-commons/django-unicorn#368] [Ideas]
Just a generic question before trying out too much.

I'd like to subclass UnicornView - to create another "template" for a component. Example: I'd like to create a component that has support for model editing, like a ModelView, but with Unicorn. I have a PolymorphicModel that has similar child models - a block on a website, each with slightly different additional fields.
I want to edit all those blocks in one View. Now I want to create a Unicorn component as base for all blocks to edit them.

Is it possible/useful to subclass UnicornView further? Like:
```python
class BlockView(UnicornView):
    title = ""
    # ...

class NewsEditor(BlockView):
    news_headline = ""
    # ...
```

is that a bad idea?

I am thinking further, and creating a "generic" UnicornView that createst its fields and their corresponding form fields dynamically.
@adamghill  Just wanted to know you opinion about that before I start something stupid that is not possible with Unicorn...

**Comment by adamghill** [comment]
>Is it possible/useful to subclass UnicornView further?
I can't think of a reason why you would run into problems subclassing as much as you want. The only issue would be if you override a method that `Unicorn` expects to be there (same as just overriding `UnicornView` itself).

>creating a "generic" UnicornView that createst its fields and their corresponding form fields dynamically
This would be super interesting and I _think_ would work as long as you created the fields at the right time. Because Python is the way it is (dynamic!), as long as it looks correct to `Unicorn` it won't be the wiser.

I look forward to hearing how this goes. 👍

**Reply by nerdoc** [reply]
I think I need some help here... I am creating a custom view using Django's builtin ModelFormMixin:
```python
class BlockEditorView(ModelFormMixin, UnicornView):
    form_class = BlockForm
    def __init__(self, **kwargs):
        ...
```

Which seems correct to me, but I always get the Error:
```
  File "/<...>/.venv/lib/python3.10/site-packages/django/views/generic/edit.py", line 44, in get_form_kwargs
    if self.request.method in ('POST', 'PUT'):
'NoneType' object has no attribute 'method'
```
which is thrown in Django's `FormMixin.get_form_kwargs()`

The Mixin accesses self.request, (which it doesn't define itself, but relies on the View that defines it later). `UnicornView` has a `request` attribute, but it seems that at the time of calling, this object is not available yet and therefore `None`.
I tried to swap `(ModelFormMixin, UnicornView)`, add `ProcessFormView`,  `TemplateView` to mess around with the MRO, etc. but nothing 
helped.

Could you enlighten me a bit - It should be correct to add a `FormMixin` to  a `UnicornView` IMHO. Is there a chance to know why this happens?
The call `if self.request.method in ('POST', 'PUT'):`is called when the component is declared IMHO, and this is too early, `self.request` is still `None` then.
Any thoughts?

**Reply by nerdoc** [reply]
Funnily, if I walk up the stack trace, the `request object is there."
* `UnicornNode.render()` has it: `rendered_component = view.render(...)`
* `UnicornView.render()`does NOT have it any more.
it seems, at a quick glance, that in [unicorn.py, Line 142](https://github.com/adamghill/django-unicorn/blob/7aa4a75a81af15374d2622678983b56cd8fde073/django_unicorn/templatetags/unicorn.py#L142), UnicornView.create() creates the view, and passes `request=request` to the view. At this time, request is still a valid HttpRequest. But it seems that that view does not save it as `self.request`... Am I correct?
At least, at [rendering time, line 155](https://github.com/adamghill/django-unicorn/blob/7aa4a75a81af15374d2622678983b56cd8fde073/django_unicorn/templatetags/unicorn.py#L155), `view.request`is `None.
Is this a Unicorn bug? I can't imagine that this is correct...

**Reply by nerdoc** [reply]
If i add a cheap `view.request = request` in [line 150](https://github.com/adamghill/django-unicorn/blob/7aa4a75a81af15374d2622678983b56cd8fde073/django_unicorn/templatetags/unicorn.py#L150), it at least works... But this is not the correct place...

**Reply by nerdoc** [reply]
Hmmm. seems that there is a problem using this views together. https://stackoverflow.com/questions/14804123/whats-the-correct-way-to-use-djangos-modelformmixin

**Reply by nerdoc** [reply]
To solve this in a really clean way, it would require to make UnicornVie not a subclass of TemplateView, but of UpdateView... And I think you don't want to do that. But then others could come and say, why not other...

* `TemplateView` (and hence `UnicornView` has `templateresponseMixin`, `ContextMixin`, and `View`as parents. 
* ` UpdateView` has `SingleOpjecttemplateresponseMixin` (which is a child of `TemplateResponseMixin`, `BaseUpdateView` (-> ProcessFormView`-> `Vew`) and `ModelFormMixin` (which is a sub-sub-child of `ContextMixin` too).

So, UpdateView "must be called with either an object pk or a slug in the URLconf" - this is not possible at all for a UnicornView which is just a component.

So I think this is not really an option here...

Sorry for the spam. But it's really hard to see through here.

**Reply by nerdoc** [reply]
I've drawn a UML diagram of Django's Template system including UnicornView:
![image](https://user-images.githubusercontent.com/2955584/155891228-6ff47eeb-deb4-4d3f-a558-4f5fad35ec6f.png)

So to use the Form mixins, it would have to be completely redesigned... and subclass e.g. UpdateView. Maybe that is the wrong approach? Or would that be a possibility? It includes `TemplateResponseMixin` as well. But it would mean that you HAVE to specify a Form in any case. Which is not always desired...

But OTOH, in most of the cases you have to use a Form in Unicorn because you want to manipulate data. Even when it's an "empty Form with a button" - meaning just a button.

but there would have to be a redesign, as said... Rebase UnicornView on FormView...

**Reply by nerdoc** [reply]
Hmmm. "TemplateView shouldn’t be used when your page has forms and does creation or update of objects. In such cases FormView, CreateView or UpdateView is a better option."

I did something brave - and just changed `class UnicornView(TemplateView)` into `class UnicornView(FormView)` in my local django_unicorn lib. And to my astonishment, it worked.
If I give the class a `form_class = MyForm`, it renders the form correctly. Even when I change `{{ form }}` to `{% crispy form %}` - everything works.

Just to have said it, *every one* of your [examples](https://www.django-unicorn.com/examples) is a form. It just would require to *use* a `Form` class.

So @adamghill, maybe it's not a too bad idea to overthink the change from TemplateView as base to FormView for UnicornView.
It would just mean to acknowledge that Unicorn **is** about forms. And if not, form_class could be `None`, or an empty dummy form.

But this could be the beginning of a complete new direction of Unicorn.

Could you tell me what you think about that?

**Reply by adamghill** [reply]
Phew! I'd have to dig into this some more!

I don't love the idea if it means that `Unicorn` requires a form. However, if it's basically hidden from the end-user and is really only for functionality with other libraries, that doesn't strike me as too bad as long as it doesn't break other uses. For example, I've used `Unicorn` as a quick and dirty component library in the past or just had links that triggered actions which didn't require a form at all.

Let me think about it a little bit and see what I can figure out!

**Reply by nerdoc** [reply]
Yes, links only as actions would not require a Form. But OTOH, inheriting from FormView would at least **let** users add a `Form`, for e.g. using CrispyForms with unicorn. ATM this is not possible at all IMHO, because TemplateView seems not to be mixable with FormView.

If there is no Form required, a linked empty Form wouldn't be a problem...

If you have another way of mixing in a `Form`/`FormMixin` into Unicorn easily, it would be the same (success).

**Reply by adamghill** [reply]
Ok, so I've never used `crispy-forms` before (!!!) so I'm probably missing something important, but I used the example form from their documentation and added an example in https://github.com/adamghill/django-unicorn/pull/372. It all seems to render as I would expect, but I'm guessing there is something else you are trying to do?

Can you take a look at my example and fork it with some code that doesn't work as you expect?

The part I don't understand without digging into their documentation more is how to add custom attributes like `unicorn:model` to the forms that get rendered. If you know how to do that, that would be awesome!

<img width="896" alt="image" src="https://user-images.githubusercontent.com/317045/156891011-2899dd6f-e839-4646-b1e3-697c4eedb8c9.png">

**Reply by nerdoc** [reply]
Ah, yes, That is not the problem. using crispy is just, as you did it in the example, surely possible. It makes no difference if you don't use crispy and put a `{{ form_class }}`  instead of  `{% crispy view form_class ... %}` in the template.
The rendering is correct. But what I want is the `FormView` behaviour: Things like `form_valid()` be called:

 *"This method is called when valid form data has been POSTed."*

in the long term I'd like to have it easy to let unicorn manage e.g. a Model's data automatically, by using a form, like ModelForm does. This would replace ModelForm, and enable editing data without reloading the page, as Unicorn does it, but for a complete Model.

I could duplicate all the things that FormView does, or, like suggested, let Unicorn inherit from FormView instead of TemplateView.
However, a `success_url` is not really a wanted aim, so "maybe" it is still anotherp ossibility to mimick FormView's behaviour.

Or am I thinking completely wrong here? And, as you said, just letting crispy etc. render the form, and processing the data after e.g. clicking a button (or another event) is a better approach...
What do you say...?

But maybe I am thinking too complicated. 

> The part I don't understand without digging into their documentation more is how to add custom attributes like `unicorn:model` to the forms that get rendered. If you know how to do that, that would be awesome!

Yes, I know how to add `unicorn:model` to that fields. I did that in my own code and it worked for an even dynamically generated form (fields created at runtime).
That's the problem, just having `form_class` doesn't make the fields "live" in a Unicorn way. You have to generate the classes fields in the `__init__` method, like this (untested, I modified it):
```python
class SomeView(UnicornView):
    form_class = ExampleForm
    def __init__(self,  **kwargs):
        for field_name in self.form_class.fields.values(): #  not sure
            # set the classes Unicorn attrs dynamically
            setattr(SomeView, field_name, "")

        super().__init__(**kwargs)

    def mount(self):
        """Loads initial data from e.g. database at component mounting"""
        for field_name in self.form_class.fields:
            # load values from database etc.
                        setattr(self, field_name, value)
```

This should work somehow for getting Unicorn attrs into that component that have the same name as the corresponding field.
Now we need to match them to a Unicorn model. We do that in the Form

```python
class ExampleForm(Form):
    def __init__(self, *args, **kwargs):
        super().__init__(*args, **kwargs)
        for field in self.fields.values():
            field.widget.attrs["unicorn:model.lazy"] = field.name  # untested
```

This is untested, but should work more or less, with a few modifications...

This is for matching the form fields fo Unicorn component attrs. What is needed at the end is button that calls a `save()` method of the component, which  saves the data into the model.

**Reply by nerdoc** [reply]
Ah, and another issue with your example... I tried it with a normal component (no direct view). When you assign `form_class = ExampleForm` to the component class, the Form is not added to the context - Could it be that only "simple" python types are added to the template context? See [here](https://github.com/nerdoc/django-unicorn-forms-test) my simple example test project - I get an error:
```
Failed lookup for key [form] in [{'True': True, 'False': False, 'None': None}, {}, {}, {'view': <app.views.MainView object at 0x7fa178386c20>,
'csrf_token': <SimpleLazyObject: 'XOizPv8bUgfvlPhyAhsJHoOPANerAGSeQnCch1GdE9dkeGYfIe43UJv0xlOTPdx3'>, 'request':
 <WSGIRequest: GET '/'>, 'user': <SimpleLazyObject: <django.contrib.auth.models.AnonymousUser object at 0x7fa170dc5f90>>, 'perms':
 PermWrapper(<SimpleLazyObject: <django.contrib.auth.models.AnonymousUser object at 0x7fa170dc5f90>>), 'messages':
 <FallbackStorage: request=<WSGIRequest: GET '/'>>, 'DEFAULT_MESSAGE_LEVELS': {'DEBUG': 10, 'INFO': 20, 'SUCCESS': 25,
 'WARNING': 30, 'ERROR': 40}, 'True': True, 'False': False, 'None': None, 'head': <bound method TemplateView.get of
<app.components.test_component.TestComponentView object at 0x7fa1783b1c90>>, 'unicorn': {'errors': {}}}]
```

This is maybe  a complete stupid error on my side, or... ?!?

**Reply by nerdoc** [reply]
Ah, I got the solution, thanks to @nielsonsantana in [another issue](https://github.com/adamghill/django-unicorn/issues/165#issuecomment-884738397). If you don't use a direct view, where you can use `{{ form }}` (or `{% crispy form %}`directly, but include a Unicorn component with a form argument, and *that* renders the form, then you have to exclude the form from javascript in the parent form.

I wrote a small chunk of code that *kind of* works in my application well until now, still exploring further:

```python
class UnicornFormMixin:
    form_class: Form = None

    class Meta:
        javascript_exclude = ("form_class",)

    def __init__(self, **kwargs):
        for field_name, field in self.form_class.fields.items():
            # set the classes Unicorn attrs dynamically
            setattr(self.__class__, field_name, "")
            field.widget.attrs["unicorn:model.lazy"] = field_name
        super().__init__(**kwargs)
```

You can "mix" that into a UnicornView:
```python
# block_edit.py

#...
all_forms = {
    "block": BlockForm,
    "header": HeaderForm,
    "map": MapForm,
    # ...
}

class BlockEditView(UnicornFormMixin, UnicornView):
    form_class: Form = None

    def __init__(self, **kwargs):
        self.form_class = all_forms[kwargs["name"]]()
        super().__init__(**kwargs)

    def updated(self, name, value):
        print(name, value)

```
So you can use the component in  a template:
```django
{% for block in ... %}
  ...
  {% unicorn block_edit name=block.name %}
  ...
{% endfor %}
```
The block component template (block_edit.html)  is just:
```django
{% load crispy_forms_tags %}
<div>
  {% crispy form_class %}
</div>
```
@adamghill , it may help in your question [how to use crispy forms](https://github.com/adamghill/django-unicorn/discussions/227) too.

It's not perfect, and I don't know the internals of Unicorn too much, so I stumbled upon the `_get_form()` method - you already have some *kind of* form support there.

But I thought a Mixin would make it a bit more modular.

Please tell me what you think about it. ATM, only the `unicorn:model`s  are attached, but that could be extended to many other things too, and made more generic.

This would be a killer feature. Using Forms with Unicorn, matching the Form's fields to Unicorn's attributes.

**Reply by nerdoc** [reply]
@adamghill , if you could give me a hint here, on Saturday I have some spare time to work on this feature a bit - but I'm stuck ATM. As stated in my [comment](https://github.com/adamghill/django-unicorn/pull/372#pullrequestreview-905976962), using a component that is included using `{% unicorn ... %}` does not work - and I can't figure out why. I always get a `VariableDoesNotExist` error, because the `form_class` variable is not correctly passed to the template context. Funnily, your example's `'name': 'test crispy'` attr is passed, and every other attr I create too. But the `form_class` attr never shows up in the template context.

It's frustrating. And I sense that I'm missing something...

**Reply by adamghill** [reply]
Hey @nerdoc sorry for the frustration! Instead of `{% crispy form_class %}` can you try `{% crispy view.form_class %}`? `view` is in the context and should give you access to the `form_class` as you need.

**Reply by nerdoc** [reply]
Nope... view is in the context, yes. but `view` in my case is the normal TemplateView (or any view where I put the Unicorn component in. `view` is not a DirectView. `form_class` is an attr of the component. So there is no `view.form_class`... That's the problem. But I can#t understand why I put a `foo = "foo"` into the component, and `foo` is in the context and can be rendered, but when I put `form_class = BlockForm` into the component, it is not passed to the context. I can't even find any code that rmoves it from context. But it is not even in `_attribute_names` ... I'm clueless.

**Reply by nerdoc** [reply]
@adamghill Please have a look at my [minimal Unicorn-Crispy example](https://github.com/nerdoc/django-unicorn-forms-test). You can just copy/paste the install instructions and should have a "not working" app in 1 minute. Not working, as I reproduce exactly this error.

**Reply by adamghill** [reply]
The reason `form_class` is not available is because of https://github.com/adamghill/django-unicorn/blob/main/django_unicorn/components/unicorn_view.py#L546 and classes _are_ considered callable in Python.

I don't love this solution and I think I could provide better developer DX to do what you are looking for, but [this PR](https://github.com/nerdoc/django-unicorn-forms-test/pull/1) has code that should at least give you a way forward. If you run into issues with this or would like a different API let me know. 👍

**Reply by nerdoc** [reply]
Thanks to @adamghill, I managed to solve this problem now. Forms have to be instantiated, and I learned much about Forms and Unicorn here.

So, I created a Mixin, first by mimicking many of Django's `FormMixin` - but then saw that it is even better to keep DRY and reuse it with a few mods. This class can be mixed into any `UnicornView`, and a `form_class` pointing to any `Form` (even `ModelForm`) added creates a "live" component of that form. This is **SO** great, and I'm quite happy with it - even if it has still a few glitches. So Please tell me what you think of that.

```python
class UnicornFormMixin(FormMixin):
    """Mixin for UnicornView to add Forms functionality."""

    def __init__(self, **kwargs):
        if self.success_url is not None:
            raise ImproperlyConfigured(
                f"You may not use a success URL attribute with Unicorn's {self.__class__.__name__}."
            )
        if self.initial is not None:
            raise ImproperlyConfigured(
                f"Do not use the 'initial' attr for setting initial data in a component. Set attributes directly in the component class '{self.__class__.__name__}' instead ."
            )

        for field_name, field in self.form_class.base_fields.items():  # type: str,Field
            # set the classes Unicorn attrs dynamically, but don't override existing "initial" values
            if not hasattr(self, field_name):
                setattr(self.__class__, field_name, "")
            # This part could be improved: Shoudl be made configureable by the user, and maybe per field?
            field.widget.attrs.update({"unicorn:model.lazy": field_name})

        super().__init__(**kwargs)

    def get_initial(self):
        return self._attributes()
```

**Reply by nerdoc** [reply]
There are some issues with e.g. in UniCornView's [get_frontend_context_variables(()](https://github.com/adamghill/django-unicorn/blob/c80ebd706d4d80d50fd1b0b660cf652845e5f17f/django_unicorn/components/unicorn_view.py#L427), and [_get_form()](https://github.com/adamghill/django-unicorn/blob/c80ebd706d4d80d50fd1b0b660cf652845e5f17f/django_unicorn/components/unicorn_view.py#L463).- it calls form.is_valid(), hence validating the whole form after each model update, which in the most cases may not be what you want: After every event, the form helpers show the error messages of all other forms if they do not validate correctly (not filled in).

I think the best way to go would be to make a [WIP PR](https://github.com/adamghill/django-unicorn/pull/378) with e.g. this code, and subsequently move code that interferes with forms from UnicornView to UnicornFormMixin - to prevent interdependencies and better encapsulation.

**Reply by adamghill** [reply]
>move code that interferes with forms from UnicornView to UnicornFormMixin - to prevent interdependencies and better encapsulation

Yes! I totally agree. Thanks for making the initial PR, let me know if you are going to keep adding to it or if I should jump in there or how you want to approach this.

For this part of your code:

```python
field.widget.attrs.update({"unicorn:model.lazy": field_name})
```

I was thinking it might make sense to have a dictionary in `Meta` that could be set that would `{'field_name': 'unicorn_attribute'}`. Something like:

```python
class SomeFormView(UnicornFormMixin, UnicornView):
    name: str

    class Meta:
        form_attributes = { 'name': 'unicorn:model' }
```

But, there also might be something that could be done in the `Form` itself? Maybe with widgets? It's been a while since I read through the forms code, though so I'm not sure.

>it calls form.is_valid(), hence validating the whole form after each model update

I know when I built the validation functionality into `Unicorn` there are a lot of tricky edge conditions and functionality that I was trying to get correct and for some reason I think I _had_ to call `is_valid` where I did. But, as part of these form changes I can re-visit all this code and see how to make it better.

**Reply by nerdoc** [reply]
> I was thinking it might make sense to have a dictionary in `Meta` that could be set that would `{'field_name': 'unicorn_attribute'}`. Something like:
> 
> ```python
> class SomeFormView(UnicornFormMixin, UnicornView):
>     name: str
> 
>     class Meta:
>         form_attributes = { 'name': 'unicorn:model' }
> ```
Yes, I thought about tht too - the code I wrote in the PR at this place is just a stub - First I thought about a dictionary too, but then I experimented with different needs, and found out that sometimes you need to use `.lazy` when on the field named `something_input`, and then, you need a `.defer` when you are in a combobox.
So IMHO the best way to keep it flexible is a function like:
```python
            field.widget.attrs.update(self.get_widget_attr(field_name, field))
```

That is called instead (maybe you find a better name?), and that flexibly returns the attrs, I updated the code [here](https://github.com/nerdoc/django-unicorn/commit/1508813a0077423f7f627709d2dad70e34d0206c) - this should be a bit better.
 
> But, there also might be something that could be done in the `Form` itself? Maybe with widgets? It's been a while since I read through the forms code, though so I'm not sure.

Yes, the form can change the widgets'  attrs too - but it makes sense to keep that in groups as well (for all BooleanFields e.g.) - this is better done in the view class.

So - you can jump in from here on - I don't have much time next week anyhow for coding I'm afraid. :fearful:

**Reply by nerdoc** [reply]
ATM this code is completely broken. Adam, do you have some time to discuss this a bit further? UnicornView takes `self.form_class` to validate the class attribute fields against. But IMHO this is a very unclean API, and does not help further with forms, because you have to add all the fields anyway to the UnicornView, and to the form. Which does not work at all when using a ModelForm - which is one of the greatest advantages of Django.
The current docs at https://www.django-unicorn.com/docs/validation/ are very confusing.

I would suggest to concentrate on having an API for Unicorn that works with any form, so ModelForm too:

```python
# models.py
class Book(models.Model):
    title = models.CharField(max_length=100)
    publish_date = models.DateField()

```
```python
# book_form.py
from django_unicorn.components import UnicornView
from django import forms
from .models import Book

class BookForm(forms.ModelForm):
    class Meta:
        model = Book
        fields = [__all__]

class BookView(UnicornView):  # or even UnicornFormView?
    book = Book.objects.none()
    form_class = BookForm

    def save():
        
```
```django
<!-- book_form.html -->
{{ form }}
{# or {% crispy form % } #}
<button u:click="save()">Save</button>
```

This is not correct, and not working, but just a rough idea.
* [x] get the form into the context: Easy done with get_context_data(), `context["form"]=self.form_class(instance=self.book)` - then the form gets correctly rendered.
* [x]  get the `unicorn:model` attrs into the fields: using `field.widget.attrs["unicorn:model"] = "<field-name>"`
* [ ] define how a save method should be called
  * normal legacy submit with preventdefault?
  * Just a custom `save()`?
  * `form_valid()` like in `ModelView`s?
* [ ] how does get_context_data know about the model? name it `model` like in `ModelView`s?
* [ ] standard way for validation

Please tell me what you think about that.

**Reply by iltoga** [reply]
The fact that unicorn doesn't play well with ModelForm is the main reason why I am marginally using it in my applications and I believe it would be a game changer for the project to have it integrated.
A very useful stack would be: unicorn View ('controller' style component) -> ModelForm ( field definition and validation) --> template (render with {{ form|crispy}})

**Reply by nerdoc** [reply]
I can only second that. I always have a look back at Unicorn (I moved to htmx meanwhile) just to see if there has anything changed at this topic. Forms are the spine of many applications, and to write your html form by hand is a task you never want to do again if you have ever worked with crispy. It would be super easy to tweak crispy to add the `unicorn:model` attribute to every form field automatically - **but** Unicorn has to implement the form rendering i the first place for components.
&lt;hint&gt;@adamghill This would be really a fast forward step for Django-Unicorn...&lt;/hint&gt;

**Reply by nerdoc** [reply]
Any progress in ModelForms? I saw that forms support has evolved somehow, but is using ModelForms possible? In a quick test I couldn't use it properly.
